students/MzKhan/lesson05/mailroom3.py

```diff
@@ -0,0 +1,177 @@
+'''
+Name: Muhammad Khan
+Date: 03/15/2019
+Assignment05
+'''
+
+import sys
+import os
+import datetime
+from operator import itemgetter as igetter
+
+#Goal:
+# Automate the mailroom job of an individual working at a local
+# charity organization. The script will prompt the user for the varioius options
+# and perform the desired tasks.
+
+# Global donors data.
+# Accessible everywhere in the script.
+donors_name = ["Adam Johnson","Matt Marvin","Ashley Wiggins",
+                "Kristina Laughrey","Kimberley Allen", "Doug Boolinger",
+                "Sherry Henning"]
+donations = [ [6000.20, 2200, 1050,100],
+                [550,1000,250],
+                [55.66,270,1000],
+                [500, 500,200],
+                [12000.99, 5000],
+                [260.57, 930.90],
+                [6000, 2460.20, 900] ]
+
+# Create the dictionary using the dict comprehension.
+donors_data = dict(zip(donors_name,donations))
+
+def prompt_user():
+    """
+    prompt the user for the valid input.
+    input: string
+    return: int
+    """
+    print('Please choose one of the following options: \n')
+    print("1: Send a Thank You")
+    print("2: Create a report")
+    print("3: Send letters to everyone")
+    print("4: Quit")
+    return int(input("Option:  "))
+
+def thank_you():
+    """send a thank you note to a donor for his/her donation."""
+    name = name_prompt()
+    amount = donation_prompt(name)
+    add_donation(name, amount)
+    email(name, amount)
+
+def add_donation(name, amount):
+    """
+    parm: string
+    parm: float
+    """
+    donors_data.setdefault(name,[]).append(amount)
+
+def name_prompt():
+    """return a name string"""
+    donor_name = input("Please enter the donor's full name OR \n"
+                 +"type 'list' for the existing donors: ")
+
+    if not "".join(donor_name.split()).isalpha():
+        raise Exception("The name should be all Alphabetic characters: ")
+    if donor_name == 'list':
+        display_list()
+        donor_name = input("Please enter the donor's full name: ")
+        if not "".join(donor_name.split()).isalpha():
+            raise Exception("The name should be all Alphabetic characters: ")
+
+    return donor_name.title()
+
+def donation_prompt(name):
+    """
+    prompt the user for the donation amount for the given donor.
+    parm: string
+    return: float
+    """
+    donation_amount = input("Please enter the donation amount for {}: "
+                                                               .format(name))
+    return float(donation_amount)
```

Comment:
what if they enter as a string? will your program crash?

Reply:
This exception will be finally "catch" in the main function. I asked the same question in the last comments. Should I catch exception in this function? Or just catch in the main function?

Reply:
The answer may vary depending on the program and kind of code it is.

In the above case you want catch the exception in here, for reasons:
* exceptions should be around the code that throws it
** you should be clear and understand what code may throw what error
* you may want to let user know with explicit message: "Please enter a number!" - right now you don't get the granularity
* you can correct user mistake instead of just letting your program falling through and the user not knowing what they did wrong.